software/control/filterwheel.py

```diff
@@ -0,0 +1,64 @@
+import os
+import time
+
+from control._def import *
+
+class SquidFilterWheelWrapper:
+
+    def __init__(self, navigationController = None):
+        # emission filter position
+        self.w_pos_index = 0
+
+        self.navigationController = navigationController
+
+    def homing(self):
+        # make sure the start homing position is not at homing position  
+        starttime = time.time() 
+        self.navigationController.home_w()
+        self.navigationController.wait_till_operation_is_completed()
+        # judge whether the current position is a homing position or not
+        if time.time() - starttime > 0.02:
+            # move offset, could be modified later
+            self.navigationController.move_w(SQUID_FILTERWHEEL_OFFSET)
+
+        self.w_pos_index = 0
+        
+    def next_position(self):
+        if self.w_pos_index < SQUID_FILTERWHEEL_MAX_INDEX:
+            self.navigationController.move_w(SCREW_PITCH_W_MM / 8.0)
+            self.navigationController.wait_till_operation_is_completed()
+            self.w_pos_index += 1 
+
+    def previous_position(self):
+        if self.w_pos_index > SQUID_FILTERWHEEL_MIN_INDEX:
+            self.navigationController.move_w(-(SCREW_PITCH_W_MM / 8.0))
+            self.navigationController.wait_till_operation_is_completed()
+            self.w_pos_index -= 1
+
+    def set_emission(self, pos):
+        if pos in range(SQUID_FILTERWHEEL_MIN_INDEX, SQUID_FILTERWHEEL_MAX_INDEX + 1): 
+            if pos != self.w_pos_index:
+                self.navigationController.move_w((pos - self.w_pos_index) * SCREW_PITCH_W_MM / 8.0)
```

Comment:
Instead of relative moves, it might be safer to use `move_w_to` with absolute moves everywhere.  That way you just keep track of `self.w_pos_index` for the `previous_position` and `next_position`, and here in `set_emission` you could just calculate the absolute position for `pos` directly and `move_w_to`.

Using absolute moves also means that if someone bumps the filter, or manually moves it, or whatever, it'll get corrected when the next filter change happens.

Reply:
@veerwang

Reply:
OK, I would add move_w_to function next time.